src/core/internal/array/construction.d

```diff
@@ -51,10 +52,17 @@ Tarr _d_arrayctor(Tarr : T[], T)(return scope Tarr to, scope Tarr from) @trusted
     {
         for (i = 0; i < to.length; i++)
         {
-            auto elem = cast(Unqual!T*)&to[i];
-            // Copy construction is defined as bit copy followed by postblit.
-            memcpy(elem, &from[i], element_size);
-            postblitRecurse(*elem);
+            static if (__traits(hasCopyConstructor, T))
+            {
+                to[i].__ctor(from[i]);
```

Comment:
This won't work for a T = static array of structs with copy ctors (the trait returns true).

Reply:
Can you drop a unittest example, please?

Reply:
Slightly modifying one of your existing ones:
```D
S[2][2] arr1;
S[2][2] arr2 = [[S(0), S(1)], [S(2), S(3)]];
_d_arrayctor(arr1[], arr2[]);
```

Reply:
So you're suggesting that the code should be changed to
```
try
    {
        static if (__traits(hasPostblit, T))
        {
            memcpy(to, from, element_size * to.length);
            for (i = 0; i < to.length; i++)
            {
                auto elem = cast(Unqual!T*)&to[i];
                postblitRecurse(*elem);
            }
        }
        else static if (__traits(hasCopyConstructor, T))
        {
            for (i = 0; i < to.length; i++)
            {
                to[i].__ctor(from[i]);
            }
        }
        else
        {
            memcpy(to, from, element_size * to.length);
        }
    catch (Exception o)
    {
        /* Destroy, in reverse order, what we've constructed so far
        */
        static if (__traits(hasPostblit, T) || !__traits(hasCopyConstructor, T))
        {
            for (int j = i; j < to.length; j++)
            {
                auto elem = cast(Unqual!T*)&to[i];
                memcpy(elem, &T.init, element_size);
            }
        }
        while (i--)
        {
            auto elem = cast(Unqual!T*)&to[i];
            destroy(*elem);
        }
        throw o;
    }
```

Reply:
Thanks for catching this. Now it's fixed